src/deepsparse/timing/timing_builder.py

```diff
@@ -0,0 +1,119 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+from typing import Dict
+
+from pydantic import BaseModel
+
+from deepsparse.timing.timing_schema import InferenceTimingSchema
+
+
+__all__ = ["TimingBuilder"]
+
+
+class TimingBuilder:
+    """
+    This object aggregates the durations
+    (time deltas in seconds) of various components of the inference
+    pipeline.
+    Once all the desired time deltas are aggregated, the
+    object may build the data structure that summarizes the
+    collected information.
+
+    Example flow:
+
+    ```
+    builder = TimingBuilder()
+
+    builder.start("total_inference")
+
+    builder.start("pre_process")
+    do_something()
+    builder.stop("pre_process")
+
+    builder.start("engine_forward")
+    do_something()
+    builder.stop("engine_forward")
+
+    builder.start("post_process")
+    do_something()
+    builder.stop("post_process")
+
+    builder.stop("total_inference")
+    ...
+    summary = builder.build()
+    ```
+    The object may time the duration of an arbitrary number
+    of events (phases), but the names of phases need to
+    be consistent with the `schema` argument passed to
+    the build() method
+    """
+
+    def __init__(self):
+        self._start_times = {}
+        self._stop_times = {}
+
+    def start(self, phase_name: str):
+        """
+        Collect the starting time of the phase
+
+        :param phase_name: The name of an event (phase), which duration
+            we are measuring
+        """
+        if phase_name in self._start_times:
+            raise ValueError(
+                f"Attempting to overwrite the start time of the phase: {phase_name}"
+            )
+        self._start_times[phase_name] = time.perf_counter()
+
+    def stop(self, phase_name: str):
+        """
+        Collect the finish time of the phase
+
+        :param phase_name: The name of an event (phase), which duration
+            we are measuring
+        """
+        if phase_name not in self._start_times:
+            raise ValueError(
+                f"Attempting to grab the stop time of the phase: {phase_name},"
+                f"but its start time is missing"
+            )
+        if phase_name in self._stop_times:
+            raise ValueError(
+                f"Attempting to overwrite the stop time of the phase: {phase_name}"
+            )
+        self._stop_times[phase_name] = time.perf_counter()
+
+    def build(self, schema: BaseModel = InferenceTimingSchema) -> BaseModel:
```

Comment:
Can we just return InferenceTimingSchema here instead of passing in a schema?

Reply:
Alternatively going back to dictionary comment - this class could just return the delta dictionary, and then whatever class is using the TimingBuilder could build it's schema using the dictionary. That way this TimingBuilder class doesn't have to know anything about schemas, but we are still capturing that type checking outside of this class

Reply:
Why don't you like the current state of the `build` method?

I could agree on returning the `InferenceTimingSchema` and not passing a `schema` as an arg.
We will be effectively limiting the flexibility of the object (you cannot name your phases however you like), but maybe for a good cause.
And then if we want to extend the usability of the TimingBuilder in the future, we can do it (and then perhaps return dicts as you proposed).

Reply:
I think because this method is doing some validating on fields you used earlier, whereas the start/stop methods don't do that validation. So there is a little bit of inconsistency.

The phase names and therefore fields of the schema are controlled outside of this class, so in my head it makes sense for building the schema to be outside of this class as well.

Agree that returning InferenceTimingSchema from this function does seem too limiting with that in mind (since the phase names are not required to be valid for InferenceTimingSchema).

Reply:
Ok, now I am leaning towards returning the dict and putting it into the schema outside of the class methods.

Reply:
agreed, I like this flow - it doesn't intrude on anything we'll need in the server and keeps the builder as something we can use generically for other timings